src/output.c

```diff
@@ -1092,37 +1196,119 @@ void OutputRegisterLoggers(void)
     /* KRB5 JSON logger. */
-    JsonKRB5LogRegister();
+    OutputRegisterTxSubModule(LOGGER_JSON_TX, "eve-log", "JsonKRB5Log", "eve-log.krb5",
+            OutputKRB5LogInitSub, ALPROTO_KRB5, JsonGenericLogger, JsonLogThreadInit,
+            JsonLogThreadDeinit, NULL);
+
+    SCLogDebug("KRB5 JSON logger registered.");
     /* QUIC JSON logger. */
-    JsonQuicLogRegister();
+    OutputRegisterTxSubModule(LOGGER_JSON_TX, "eve-log", "JsonQuicLog", "eve-log.quic",
+            OutputQuicLogInitSub, ALPROTO_QUIC, JsonGenericLogger, JsonLogThreadInit,
+            JsonLogThreadDeinit, NULL);
+
+    SCLogDebug("quic json logger registered.");
     /* DHCP JSON logger. */
     JsonDHCPLogRegister();
     /* SNMP JSON logger. */
-    JsonSNMPLogRegister();
+    OutputRegisterTxSubModule(LOGGER_JSON_TX, "eve-log", "JsonSNMPLog", "eve-log.snmp",
+            OutputSNMPLogInitSub, ALPROTO_SNMP, JsonGenericLogger, JsonLogThreadInit,
+            JsonLogThreadDeinit, NULL);
+
+    SCLogDebug("SNMP JSON logger registered.");
     /* SIP JSON logger. */
-    JsonSIPLogRegister();
+    OutputRegisterTxSubModule(LOGGER_JSON_TX, "eve-log", "JsonSIPLog", "eve-log.sip",
+            OutputSIPLogInitSub, ALPROTO_SIP, JsonGenericLogger, JsonLogThreadInit,
+            JsonLogThreadDeinit, NULL);
+
+    SCLogDebug("SIP JSON logger registered.");
     /* RFB JSON logger. */
-    JsonRFBLogRegister();
+    OutputRegisterTxSubModule(LOGGER_JSON_TX, "eve-log", "JsonRFBLog", "eve-log.rfb",
+            OutputRFBLogInitSub, ALPROTO_RFB, JsonGenericLogger, JsonLogThreadInit,
+            JsonLogThreadDeinit, NULL);
     /* MQTT JSON logger. */
     JsonMQTTLogRegister();
     /* Pgsql JSON logger. */
     JsonPgsqlLogRegister();
     /* Template JSON logger. */
-    JsonTemplateLogRegister();
+    OutputRegisterTxSubModule(LOGGER_JSON_TX, "eve-log", "JsonTemplateLog", "eve-log.template",
+            OutputTemplateLogInitSub, ALPROTO_TEMPLATE, JsonGenericLogger, JsonLogThreadInit,
+            JsonLogThreadDeinit, NULL);
     /* RDP JSON logger. */
-    JsonRdpLogRegister();
+    OutputRegisterTxSubModule(LOGGER_JSON_TX, "eve-log", "JsonRdpLog", "eve-log.rdp",
+            OutputRdpLogInitSub, ALPROTO_RDP, JsonGenericLogger, JsonLogThreadInit,
+            JsonLogThreadDeinit, NULL);
+    SCLogDebug("rdp json logger registered.");
     /* DCERPC JSON logger. */
     JsonDCERPCLogRegister();
     /* app layer frames */
     JsonFrameLogRegister();
     /* BitTorrent DHT JSON logger */
-    JsonBitTorrentDHTLogRegister();
+    if (ConfGetNode("app-layer.protocols.bittorrent-dht") != NULL) {
+        /* Register as an eve sub-module. */
+        OutputRegisterTxSubModule(LOGGER_JSON_TX, "eve-log", "JsonBitTorrentDHTLog",
+                "eve-log.bittorrent-dht", OutputBitTorrentDHTLogInitSub, ALPROTO_BITTORRENT_DHT,
+                JsonGenericLogger, JsonLogThreadInit, JsonLogThreadDeinit, NULL);
+    }
+}
+
+static AppLayerLogger alert_applayer_loggers[ALPROTO_MAX] = {
```

Comment:
I don't really like this idea of handling these per protocol things be in a central table. Seems an anti-pattern wrt trying to make it as much a plugin as possible.

Reply:
The use of this central table is indeed the central point, and it is meant to allow the use of app-layer plugins

What do we want ?
I think we want
- to avoid duplicating boilerplate code (and remove C file, to get better rust stats)
- that every app-layer protocols log their metadata for alerts, and for their own events

The solution currently used is to have
1.  For alerts, output-json-alert.c has a big switch with boilerplate wrapper code calling hardcoded app-layer logging functions, and this is missing for several protocols

So, the use of a central table, for alerts, 
- makes it easy to sport the missing protocols (and to add them, as is done in this PR for krb5 for instance)
- removes the boilerplate code (taking the mark, trying to log, closing or restoring the mark...)
- makes it easy to extend by plugins (that "just" need to extend the array)

2. For app-layer events, we have registration of boilerplate-wrapper functions that are registered in a central table
 `static OutputTxLogger *list[ALPROTO_MAX]` in output-tx.c 

Here, this PR 
- removes boilerplate code for simple logger (not ones using state, or flags...), like `JsonLogThreadInit` already exists but is not used by every protocol and `JsonGenericLogger` is introduced making use of the new table for simple loggers

So @victorjulien what do you think ?
It looks like a design flaw to have different wrapper functions for logging alert and app-layer metadata (causing not to have app-layer metadata for every protocol/alert)
Most of the protocols/cases are simple, but some require more complexity

Should I make a first part of this PR with the uncontroversial changes ? (first commits, use of `JsonLogThreadInit`)

Reply:
I think a centralized store doesn't have to use a table of hardcoded entries, but could also be more dynamic, like with many other parts of the engine. The parsers could then all a registration function that adds their entry into the store.

Reply:
Well, not sure to see the point but this is quite easy to add :-)